### PR TITLE
normalization should not change KeyDataPair to NameValuePair

### LINKED ISSUE #532: RequestParameter normalization replaces KeyDataPair with NameValuePair
Request Parameters of type KeyDataPair are replaced by instances of type NameValuePair when WebRequest#getParameters is called.
This introduces issues with Spring Boot 3 and especially spring-test 6 where the getParameters method is called instead of getRequestParameters as it has been in spring-test 5.x

See https://github.com/HtmlUnit/htmlunit/commit/39c58d2b114bebb157be57d63416f629630eee07#r93713845

I've already a proposal how this can be fixed. Pull Request incoming :-D

### PR DESCRIPTION
**Pull request by thuri**
Fixes #532 
See https://github.com/HtmlUnit/htmlunit/commit/39c58d2b114bebb157be57d63416f629630eee07#r93713845

### CODE REVIEW

File: src/main/java/com/gargoylesoftware/htmlunit/util/NameValuePair.java [hunk 1]
```diff
@@ -90,4 +90,8 @@ public int hashCode() {
     public String toString() {
         return name_ + "=" + value_;
     }
+
+    public NameValuePair normalized() {
```
Comment (rbri) [inline src/main/java/com/gargoylesoftware/htmlunit/util/NameValuePair.java:101]: can you please add the jdoc for this method including the INTERNAL API hint

File: src/test/java/com/gargoylesoftware/htmlunit/WebRequestTest.java [hunk 2]
```diff
@@ -425,4 +428,19 @@ public void getRequestParametersFromUrlEncodedBodyPost() throws Exception {
 
         assertEquals(0, request.getRequestParameters().size());
     }
+
+    @Test
```
Comment (rbri) [inline src/test/java/com/gargoylesoftware/htmlunit/WebRequestTest.java:432]: i like to have a bit more detailed test - checking the values of the new normalized version
and maybe also testing the normalization - having a parameter with null here or two separate tests

**Review by rbri — CHANGES_REQUESTED** [review]
Looks reasonable for me, thanks for this.
Please add yourself as author to all changed files and maybe also to the pom (if you are not already in the list)

File: src/main/java/com/gargoylesoftware/htmlunit/util/NameValuePair.java (hunk 1 above)
Comment (thuri) [reply src/main/java/com/gargoylesoftware/htmlunit/util/NameValuePair.java:101]: done. Squashed and force pushed to PR source branch

File: src/test/java/com/gargoylesoftware/htmlunit/WebRequestTest.java (hunk 2 above)
Comment (thuri) [reply src/test/java/com/gargoylesoftware/htmlunit/WebRequestTest.java:432]: I elaborated a bit on the tests. Please have a look at the new commit. If you think i should add some more testing please let me know.

**Comment by rbri** [comment]
Jenkins will have a look at this....

**Comment by rbri** [comment]
sorry we are still at java8 :-)

    /var/lib/jenkins/workspace/htmlunit_PR-533/src/test/java/com/gargoylesoftware/htmlunit/WebRequestTest.java:[437,42] cannot find symbol
    symbol:   method of(com.gargoylesoftware.htmlunit.util.NameValuePair,com.gargoylesoftware.htmlunit.util.KeyDataPair)
    location: interface java.util.List

File: src/test/java/com/gargoylesoftware/htmlunit/WebRequestTest.java (hunk 2 above)
Comment (rbri) [reply src/test/java/com/gargoylesoftware/htmlunit/WebRequestTest.java:432]: fine for me

**Comment by thuri** [comment]
Yeah I've seen that but somehow eclipse didn't warn me when i used Java 11 features.
I corrected the problematic part. At least I hope so

File: src/main/java/com/gargoylesoftware/htmlunit/util/KeyDataPair.java [hunk 3]
```diff
@@ -159,4 +160,25 @@ private void readObject(final ObjectInputStream ois) throws ClassNotFoundExcepti
...
+     *
+     * specialization of inherited method which will copy all fields
+     * and make sure that the value in the base class is not null, by calling
+     * the constructor with the current value
+     */
+    @Override
+    public KeyDataPair normalized() {
+        final KeyDataPair pair = new KeyDataPair(
+            this.getName(),
+            this.fileObject_ != null ? this.fileObject_ : new File(this.getValue()),
```
Comment (rbri) [inline src/main/java/com/gargoylesoftware/htmlunit/util/KeyDataPair.java]: new File(...) might do some normalization on the file name leading in the end to some new value -> Testcase?
maybe it is better to have a private constructor that allows to pass the values 1:1?

What do you think?

**Comment by rbri** [comment]
tests are looking good !

**Comment by thuri** [comment]
@rbri  please have a look at the additional commit. Perhaps that is what you had in mind ...

**Comment by rbri** [comment]
i have nothing in mind ;-)
thanks a lot - will inform when the new snapshot is availble

and all the best for 2023

**Comment by thuri** [comment]
Thanks a lot and the all the best for you too.
